Fix #3291: Add check for XML syntax validation

Sparsh1212 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fixes #3291: Add check for XML syntac validation

The script traverses different layers of the codebase to check for the syntax correctness of XML files
To run the script, use:

`bazel run //scripts:xml_syntax_check -- $(pwd)`

For testing the script, automated tests have been added.
To execute the tests, use:

`bazel test //scripts/src/javatests/org/oppia/android/scripts/xml:XmlSyntaxCheckTest`

**Note:** We are generating the test assets dynamically at the time of executing them. The test assets are automatically deleted, when the test finishes.

**Blocked on:** #3340 

Screenshot of new tests passing locally:

![image](https://user-images.githubusercontent.com/55937724/123278466-97f6b980-d524-11eb-8b0c-c149292fd126.png)


## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Sparsh1212 committed:
Fix #3290: Add support for generic regex pattern matching

Sparsh1212 committed:
Apply review suggestions for proto location

Sparsh1212 committed:
Delete main directory

Sparsh1212 committed:
Implment script logic and tests

Sparsh1212 committed:
CI setup part 1

Sparsh1212 committed:
syntax fix in yaml file

Sparsh1212 committed:
import nits in dummy file

Sparsh1212 committed:
Use regex patterns from script in script test

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Improve naming of script_assets variables

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Make PR suggestions

Sparsh1212 committed:
Make PR suggestions

Sparsh1212 committed:
Revamp testing approach

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Fix #3291: Add check for XML syntax correctness

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Introduce XML syntax check in static checks workflow

anandwana001 commented on .github/workflows/static_checks.yml:
```diff
@@ -107,3 +107,7 @@ jobs:
+
+      - name: XML Syntax Validation Check
+        run: |
+          bazel run //scripts:xml_syntax_check -- $(pwd) app data utility domain
```
data domain doesn't have any res file, utility does file dimen and string XML file if we are checking that else no needed, we can keep the check only for app module.

Sparsh1212 replied:
Yes! they don't have any res files, but my only concern was that these layers do have the `AndroidManifest.xml` file present, so I thought of including them too in the XML syntax check. 
WDYT?

anandwana001 replied:
Looking for @BenHenning thoughts on this

BenHenning replied:
Hmm. I just left a comment in #3340 about passing in the root directory and just recursing among that. If we validate all XML files then this slight oddity is less noticeable (since we do ultimately want all XML files to be valid). Would that mitigate the concern here?

Sparsh1212 replied:
Done.

anandwana001 reviewed (changes requested):
Thanks @Sparsh1212 

Same comment from the previous PR review, for script `kt` file and test file.

Sparsh1212 committed:
Make nit suggestions

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Make nit suggestions

Sparsh1212 commented:
> Thanks @Sparsh1212
> 
> Same comment from the previous PR review, for script `kt` file and test file.

Thanks, @anandwana001 for the suggestions. 
The PR is now updated. Please take a look.

fsharpasharp reviewed (approved):
Approving for BUILD.bazel change

Sparsh1212 committed:
Add if: always()

BenHenning commented:
Please address my comments from #3340 & apply the similar feedback to this & your next PRs, then assign this back to me. Please make sure to ensure all documentation & code is properly formatted, clear, and that you add all test cases that you can consider.

anandwana001 commented:
@Sparsh1212 Follow guide - https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section

Sparsh1212 committed:
Apply review suggestions on PR

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Make review suggestions from regex pattern checks

Sparsh1212 commented:
@anandwana001 @BenHenning The PR is now updated taking into account all the suggestions received on #3340. Please, take a look.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/XMLSyntaxCheck.kt:
```diff
@@ -0,0 +1,146 @@
+import org.xml.sax.ErrorHandler
+import org.oppia.android.scripts.ScriptResultConstants
+
+class XMLSyntaxCheck {
```
add kdocs

Sparsh1212 replied:
done

BenHenning replied:
I apparently need to reopen this since GitHub now longer allows multiple conversations for a single line of code (see: https://github.community/t/please-re-add-multiple-comments-per-line/187397).

Nit on naming: ``XmlSyntaxCheck`` per camel-casing rules: https://developer.android.com/kotlin/style-guide#camel_case

Sparsh1212 replied:
This is corrected now.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/XMLSyntaxCheck.kt:
```diff
@@ -0,0 +1,146 @@
+  companion object {
+    @JvmStatic
+    fun main(vararg args: String) {
+      /** path of the repo to be analyzed. */
```
single line comment `//`

Sparsh1212 replied:
done

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/XMLSyntaxCheck.kt:
```diff
@@ -0,0 +1,146 @@
+      docBuilder.setErrorHandler(SyntaxErrorHandler(repoPath = repoPath))
+
+      /** check if the repo has any syntactically incorrect XML. */
+      val hasXmlSyntaxFailure = searchFiles.fold(initial = false) { isFailing, file->
```
```suggestion
      val hasXmlSyntaxFailure = searchFiles.fold(initial = false) { isFailing, file ->
```

Sparsh1212 replied:
done

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/XMLSyntaxCheck.kt:
```diff
@@ -0,0 +1,146 @@
+       * @param allowedDirectories a list of all the directories which needs to be checked
+       * @return all files which needs to be checked
+       */
+      fun collectSearchFiles(
+        repoPath: String,
+        allowedDirectories: List<String>)
+        : List<File> {
```
```suggestion
      fun collectSearchFiles(
        repoPath: String,
        allowedDirectories: List<String>
      )
```

Sparsh1212 replied:
done

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/XMLSyntaxCheck.kt:
```diff
@@ -0,0 +1,146 @@
+        return File(repoPath).walk().filter { it ->
+          checkIfAllowedDirectory(
+            retrieveFilePath(it, repoPath),
+            allowedDirectories)
```
```suggestion
            allowedDirectories
          )
```

Sparsh1212 replied:
done

anandwana001 reviewed (changes requested):
Suggesting to follow my suggestion from the https://github.com/oppia/oppia-android/pull/3340 latest PR review

Sparsh1212 committed:
Refactor PR as per feedback recieved

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Implement feedback suggestions

Sparsh1212 commented:
> Suggesting to follow my suggestion from the #3340 latest PR review

Thanks, @anandwana001. 
The PR is updated on the basis of comments received on #3340 and new suggestions too. PTAL.

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Implement review suggestions

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/XMLSyntaxCheck.kt:
```diff
@@ -0,0 +1,105 @@
+      }
+    }
+
+    /** Check if a file has a broken XML syntax.
```
correct format - https://kotlinlang.org/docs/kotlin-doc.html#kdoc-syntax

Sparsh1212 replied:
Apologies! Corrected. PTAL.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/XMLSyntaxCheck.kt:
```diff
@@ -0,0 +1,105 @@
+   * Class for custom error handling of the parse exception thrown
+   * by the parser, to log the failure message in a cleaner way.
+   */
+  private class SyntaxErrorHandler(val repoPath: String) : ErrorHandler {
```
can we separate out this class?

Sparsh1212 replied:
Separated. PTAL.

Sparsh1212 committed:
nits

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Nit fixes

Sparsh1212 committed:
Nit fixes

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

anandwana001 reviewed (changes requested):
Please check on failing checks.

Sparsh1212 commented:
> Please check on failing checks.

@anandwana001 
The checks are failing to due to a bug in `compute_affected_tests.sh` in the CI workflow. There is a PR
 https://github.com/oppia/oppia-android/pull/3374 
open currently which will fix this issue. It is passing successfully on local. This applies to all the PRs: #3340 #3341 #3343 #3352

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

vinitamurthi reviewed (approved):
LGTM for codeowner files

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/SyntaxErrorHandler.kt:
```diff
@@ -0,0 +1,41 @@
+   * @param repoPath the path of the repo to be analyzed
+   * @return log the failure
+   */
+  private fun logCheckFailure(e: SAXParseException, repoPath: String) {
```
arguments name should be readable

Sparsh1212 replied:
Done. Made them more descriptive.

Sparsh1212 committed:
nit fix

anandwana001 reviewed (approved):
LGTM

BenHenning commented on .github/workflows/static_checks.yml:
```diff
@@ -107,3 +107,8 @@ jobs:
       - name: Regex Patterns Validation Check
         run: |
```
Should this also be ``if: always()`` in #3340?

Sparsh1212 replied:
Done.

BenHenning commented on scripts/BUILD.bazel:
```diff
@@ -46,3 +46,34 @@ kt_jvm_test(
+    runtime_deps = [":xml_syntax_check_lib"],
+)
+
+kt_jvm_library(
+    name = "xml_syntax_check_lib",
+    testonly = True,
+    srcs = [
+        "src/java/org/oppia/android/scripts/RepoFile.kt",
+        "src/java/org/oppia/android/scripts/ScriptResultConstants.kt",
+        "src/java/org/oppia/android/scripts/SyntaxErrorHandler.kt",
+        "src/java/org/oppia/android/scripts/XMLSyntaxCheck.kt",
+    ],
+)
+
+kt_jvm_test(
+    name = "test_xml_syntax_check",
+    testonly = True,
+    srcs = ["src/test/org/oppia/android/scripts/XMLSyntaxCheckTest.kt"],
+    test_class = "org.oppia.android.scripts.XMLSyntaxCheckTest",
+    deps = [
+        ":xml_syntax_check_lib",
+        "//testing:assertion_helpers",
+        "//third_party:com_google_truth_truth",
+        "//third_party:org_jetbrains_kotlin_kotlin-test-junit",
+    ],
+)
```
Ditto here--these should be at the file level.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/ScriptResultConstants.kt:
```diff
@@ -5,4 +5,8 @@ object ScriptResultConstants {
   val REGEX_CHECKS_FAILED: String = "REGEX PATTERN CHECKS FAILED"
+
+  /** XMLSyntaxCheck output message constants **/
+  val XML_SYNTAX_CHECK_PASSED: String = "XML SYNTAX CHECK PASSED"
+  val XML_SYNTAX_CHECK_FAILED: String = "XML SYNTAX CHECK FAILED"
```
Both of these need KDocs.

Sparsh1212 replied:
Added.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/ScriptResultConstants.kt:
```diff
@@ -5,4 +5,8 @@ object ScriptResultConstants {
   val REGEX_CHECKS_FAILED: String = "REGEX PATTERN CHECKS FAILED"
+
+  /** XMLSyntaxCheck output message constants **/
+  val XML_SYNTAX_CHECK_PASSED: String = "XML SYNTAX CHECK PASSED"
```
Maybe ``XML_SYNTAX_CHECK_PASSED_OUTPUT_INDICATOR`` might be a bit clearer on how this is being used. WDYT?

Ditto for the others.

Sparsh1212 replied:
SGTM. DONE.

BenHenning commented on scripts/BUILD.bazel:
```diff
@@ -46,3 +46,34 @@ kt_jvm_test(
+)
+
+kt_jvm_test(
+    name = "test_xml_syntax_check",
```
For this & the regex pattern: see other test targets in the codebase to ensure we naming these consistently.

This should be XmlSyntaxCheckTest (to match the name of the test).

/cc @fsharpasharp as well since tests are an intentional deviation in naming strategy so that it's easier to create grouped test commands.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/BUILD.bazel:
```diff
@@ -46,3 +46,34 @@ kt_jvm_test(
+    name = "test_xml_syntax_check",
+    testonly = True,
+    srcs = ["src/test/org/oppia/android/scripts/XMLSyntaxCheckTest.kt"],
+    test_class = "org.oppia.android.scripts.XMLSyntaxCheckTest",
```
Hmm. Could you try something: if you add an empty BUILD.bazel file in org/, org/oppia, and org/oppia/android, can you omit this test_class and still run the test?

BenHenning replied:
Following up: you can simplify this a bit (and the extra BUILD files aren't needed since I think this directory structure is better fitting what Bazel expects). See:

```bazel
kt_jvm_test(
    name = "ComputeAffectedTestsTest",
    srcs = ["ComputeAffectedTestsTest.kt"],
    deps = [
        ":compute_affected_tests_lib",
        "//third_party:com_google_truth_truth",
        "//third_party:org_jetbrains_kotlin_kotlin-test-junit",
    ],
)
```

That works for me when running locally:

```shell
bazel test //scripts/src/java/org/oppia/android/scripts/ci:ComputeAffectedTestsTest
```

Sparsh1212 replied:
Updated.

BenHenning commented on scripts/BUILD.bazel:
```diff
@@ -46,3 +46,34 @@ kt_jvm_test(
+    name = "xml_syntax_check_lib",
+    testonly = True,
+    srcs = [
+        "src/java/org/oppia/android/scripts/RepoFile.kt",
```
This should be in its own library & added as a dep so it can be shared between different scripts that need it. Ditto for any of the other generic classes (prefer using libraries over duplicating sources--this will make the build perform more efficiently).

Sparsh1212 replied:
Outdated.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/SyntaxErrorHandler.kt:
```diff
@@ -0,0 +1,41 @@
+import org.xml.sax.ErrorHandler
+import org.xml.sax.SAXParseException
+
+/**
+ * Class for custom error handling of the parse exception thrown
+ * by the parser, to log the failure message in a cleaner way.
+ */
```
Here & everywhere else: KDocs should fit the full 100 character limit before being line-wrapped just like code (prefer not prematurely line-wrapping).

Sparsh1212 replied:
Done. noted.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/SyntaxErrorHandler.kt:
```diff
@@ -0,0 +1,41 @@
+ * Class for custom error handling of the parse exception thrown
+ * by the parser, to log the failure message in a cleaner way.
+ */
+class SyntaxErrorHandler(val repoPath: String) : ErrorHandler {
```
Does ``repoPath`` need to be public? In general, prefer to keep everything with as minimal visibility as necessary to simplify your public contract (put another way: everything that's public needs to be tested thoroughly since other components can use them).

BenHenning replied:
Also, please add tests for this file (but I suggest looking at my comment in the main script class first since this class's contract could be simplified which would result in simpler tests).

Sparsh1212 replied:
Outdated. Tests added too.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/SyntaxErrorHandler.kt:
```diff
@@ -0,0 +1,41 @@
+ * by the parser, to log the failure message in a cleaner way.
+ */
+class SyntaxErrorHandler(val repoPath: String) : ErrorHandler {
+  override fun warning(e: SAXParseException) {
```
It seems that this class is specific to XML. Consider naming it: XmlSyntaxErrorHandler so that it's not confused among other error handling happening for other scripts.

This might also be a good indication that we need to do further subpackaging. What do you think about:

org.oppia.android.apps.scripts.regex
org.oppia.android.apps.scripts.xml
org.oppia.android.apps.scripts.common <-- common utilities shared between the scripts

For organizing the scripts?

Sparsh1212 replied:
SGTM. DONE.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/SyntaxErrorHandler.kt:
```diff
@@ -0,0 +1,41 @@
+   *
+   * @param e the parsing exception thrown by the parser
+   * @param repoPath the path of the repo to be analyzed
+   * @return log the failure
```
Is this correct? Nothing seems to be returned. (If the return value is Unit or Nothing then ``@return`` shouldn't be in the KDoc).

Sparsh1212 replied:
Removed. noted.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/SyntaxErrorHandler.kt:
```diff
@@ -0,0 +1,41 @@
+   * @return log the failure
+   */
+  private fun logCheckFailure(parseException: SAXParseException, repoPath: String) {
+    val failureMessage =
```
I suggest formatting this message similarly to the version suggested in #3340.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/XMLSyntaxCheck.kt:
```diff
@@ -0,0 +1,71 @@
+  companion object {
+    @JvmStatic
+    fun main(vararg args: String) {
+      // path of the repo to be analyzed.
```
Please fix single line comment casing & line wrapping per earlier comments here & in #3340.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/XMLSyntaxCheck.kt:
```diff
@@ -0,0 +1,71 @@
+   */
+  private fun checkIfFileHasBrokenXml(docBuilder: DocumentBuilder, file: File): Boolean {
+    try {
+      docBuilder.parse(file)
```
Should we maybe create a new builder for each file instead of sharing them? I'm not sure whether the builders are meant to be reused like this (or are they?).

Sparsh1212 replied:
I think it's fine to use a single builder, as we are doing nothing more than checking an Xml file can be parsed successfully or not.

Sparsh1212 replied:
Update: We are now using a new builder for every file.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/XMLSyntaxCheck.kt:
```diff
@@ -0,0 +1,71 @@
+  private fun checkIfFileHasBrokenXml(docBuilder: DocumentBuilder, file: File): Boolean {
+    try {
+      docBuilder.parse(file)
+    } catch (e: SAXParseException) {
```
I wonder if maybe a cleaner way of handling this is to collect all the errors from the error handler, and if any are present then we determine it's a failure & log them. This keeps the logging & the logic in the same file since the error handler just becomes a cataloging utility, and it avoids using exceptions for control flow (which is considered a bad practice in Effective Java).

Sparsh1212 replied:
This sounds neat to me. Here is what I am thinking, we will maintain a mutable list of errors in the `XmlSyntaxErrorHandler`, and whenever any exception is raised by the parser, we will simply add it to the list. After the parsing is done, we will just collect the list of all the errors received by the error handler and then log them and fail the script if it's the case. How does this sound?

Also, one thing that we have to take care of is that, for every `fatalError` in the Xml syntax, if the override `fatalError` method of our error handler isn't throwing any exception,  then the default error handler throws a `SaxParseException`. So, in order to make our script continue to work without being terminated in between, we have to wrap the parsing line of code in a try-catch block, just like this:
```kotlin
try {
 builder.parse('path_to_xml')
} catch (e: SaxParseException){}
```
Please let me know your thoughts on this overall approach.

BenHenning replied:
Both sound good to me, and I agree with the overall approach being proposed here.

Sparsh1212 replied:
Implemented.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/XMLSyntaxCheck.kt:
```diff
@@ -0,0 +1,71 @@
+class XMLSyntaxCheck {
+  companion object {
+    @JvmStatic
+    fun main(vararg args: String) {
```
For this & the regex check script: the main function should be documented as well. See my PR for an idea on how to do this--it's nice to document the actual usage of the script in the KDoc of main() since that's essentially how to "call" it.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/test/org/oppia/android/scripts/XMLSyntaxCheckTest.kt:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  @Test
+  fun testXmlSyntax_validXML_xmlSyntaxIsCorrect() {
```
``validXml``

Ditto elsewhere (make sure 'xml' is following expected camel case naming rules for all identifiers where it's used, including variables).

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/test/org/oppia/android/scripts/XMLSyntaxCheckTest.kt:
```diff
@@ -0,0 +1,155 @@
+  @Test
+  fun testXmlSyntax_validXML_xmlSyntaxIsCorrect() {
+    val validXML =
+      "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n" +
+        "<shape xmlns:android=\"http://schemas.android.com/apk/res/android\"\n" +
+        "  android:shape=\"rectangle\">\n" +
+        "  <solid android:color=\"#3333334D\" />\n" +
+        "  <size android:height=\"1dp\" />\n" +
+        "</shape>"
```
Maybe use ``"""`` syntax, instead? Ditto for the other tests.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/test/org/oppia/android/scripts/XMLSyntaxCheckTest.kt:
```diff
@@ -0,0 +1,155 @@
+      runScript()
+    }
+
+    assertThat(exception).hasMessageThat().contains(
+      ScriptResultConstants.XML_SYNTAX_CHECK_FAILED
+    )
```
This can fit on one line--suggest doing that here & elsewhere (including #3340).

Also, maybe directly import the constant to avoid needing to qualify it (though that may be eliminated, anyway, if we stop using the constants object).

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/test/org/oppia/android/scripts/XMLSyntaxCheckTest.kt:
```diff
@@ -0,0 +1,155 @@
+        " character data or markup.\n" +
+        "lineNumber: 4\n" +
+        "columnNumber: 4\n" +
+        "File: [ROOT]/testfiles/TestFile.xml"
```
I'm starting to change my thinking on this. Instead of ``//`` maybe we should actually just fully qualify it (e.g. absolute path).

For the tests, you can just use multiple ``contains`` checks instead of validating the exact output, and the contains can just verify ``testfiles/TestFile.xml`` is present. Ditto for other tests & #3340.

WDYT?

Sparsh1212 replied:
SGTM. DONE.

BenHenning commented on scripts/src/test/org/oppia/android/scripts/XMLSyntaxCheckTest.kt:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  @Test
+  fun testXmlSyntax_wrongClosingTag_xmlSyntaxIsInCorrect() {
```
``xmlSyntaxIsIncorrect``

Ditto elsewhere.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/test/org/oppia/android/scripts/XMLSyntaxCheckTest.kt:
```diff
@@ -0,0 +1,155 @@
+    )
+    assertThat(outContent.toString().trim()).isEqualTo(
+      "XML syntax error: The end-tag for element type \"shape\" must end" +
+        " with a '>' delimiter.\n" +
```
This error actually seems wrong--shouldn't it be mentioning that the closing tag is missing or wrong rather than missing a ``>``?

Sparsh1212 replied:
Yes, this might seem a bit incorrect, but the same error message can be reproduced on several of the online XML validators too. When we do the same syntax error, all of them report the same error message: 
"The end-tag for element type "shape" must end with a '>' delimiter."
https://www.xmlvalidation.com/index.php?id=1&L=0
https://www.liquid-technologies.com/online-xml-validator
https://www.truugo.com/xml_validator/
https://onlinexmltools.com/validate-xml

Will it be fine, to go with this? as it seems pretty difficult to dig on how we can change this, as this is reported by the parser.

Sparsh1212 replied:
> Yes, this might seem a bit incorrect, but the same error message can be reproduced on several of the online XML validators too. When we do the same syntax error, all of them report the same error message:
> "The end-tag for element type "shape" must end with a '>' delimiter."
> https://www.xmlvalidation.com/index.php?id=1&L=0
> https://www.liquid-technologies.com/online-xml-validator
> https://www.truugo.com/xml_validator/
> https://onlinexmltools.com/validate-xml
> 
> I think it might be fine to go with this.

BenHenning replied:
Yeah, I don't think we have any choice here. I just wanted to make sure that this is working as intended, even if it isn't ideal. Thanks for clarifying.

Sparsh1212 replied:
I think we can resolve this now.

BenHenning reviewed (changes requested):
Thanks @Sparsh1212! Left a bunch of thoughts. Note that a lot of these comments also apply to #3340.

Sparsh1212 committed:
Review suggestions part 1

Sparsh1212 committed:
Implement review suggestions part 2

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
resolve merge conflict

Sparsh1212 committed:
update static_checks

Sparsh1212 committed:
add test to repository file

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
merge regex-checks

Sparsh1212 committed:
Do review suggestions

Sparsh1212 commented:
> Thanks @Sparsh1212! Left a bunch of thoughts. Note that a lot of these comments also apply to #3340.

Thanks, @BenHenning.
PR is updated after all the suggestions.
Please check.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/common/ScriptResultConstants.kt:
```diff
@@ -4,3 +4,7 @@ package org.oppia.android.scripts.common
 const val REGEX_CHECK_PASSED_OUTPUT_INDICATOR: String = "REGEX PATTERN CHECKS PASSED"
 /** RegexPatternValidationCheck failed indicator. */
 const val REGEX_CHECK_FAILED_OUTPUT_INDICATOR: String = "REGEX PATTERN CHECKS FAILED"
+/** XmlSyntaxCheck successfuly passed indicator. */
```
Nit: please add spaces between the constants such that it looks like this:

```kotlin
/** KDoc...*/
constant

/** KDoc... */
constant

...
```

Sparsh1212 replied:
Done

BenHenning commented on scripts/src/java/org/oppia/android/scripts/common/ScriptResultConstants.kt:
```diff
@@ -4,3 +4,7 @@ package org.oppia.android.scripts.common
+/** XmlSyntaxCheck successfuly passed indicator. */
+const val XML_SYNTAX_CHECK_PASSED_OUTPUT_INDICATOR: String = "XML SYNTAX CHECK PASSED"
+/** XmlSyntaxCheck failed indicator. */
+const val XML_SYNTAX_CHECK_FAILED_OUTPUT_INDICATOR: String = "XML SYNTAX CHECK FAILED"
```
EOF newline here also.

Sparsh1212 replied:
Done

BenHenning commented on scripts/src/java/org/oppia/android/scripts/xml/BUILD.bazel:
```diff
@@ -0,0 +1,48 @@
+"""
+Package for Xml syntax based checks.
```
```suggestion
Package for XML syntax based checks.
```

(XML is technically capitalized when documenting it, but it should be 'Xml' when in an identifier per identifier camel-casing guidelines).

Sparsh1212 replied:
Done

BenHenning commented on scripts/src/java/org/oppia/android/scripts/xml/BUILD.bazel:
```diff
@@ -0,0 +1,48 @@
+kt_jvm_library(
+    name = "xml_syntax_error_handler_lib",
+    testonly = True,
+    srcs = ["XmlSyntaxErrorHandler.kt"],
```
Why not just include this in xml_syntax_check_lib? Will it ever be used anywhere else?

Sparsh1212 replied:
Ah! this lib is also used in `XmlSyntaxErrorHandlerTest`, so that's why I decided to separate it out.

BenHenning replied:
Got it. In that case, let's strike "_lib" from the library name since it's not needed here. We only use that for the root script libraries since they have binary targets with the same name which could introduce confusion in the build graph. By default, we don't include "_lib" since it's redundant.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/xml/XmlSyntaxCheck.kt:
```diff
@@ -0,0 +1,85 @@
+import javax.xml.parsers.DocumentBuilderFactory
+
+/**
+ * Script for ensuring that all Xml files in the repo are syntactically correct.
```
Ditto here & elsewhere for XML.

Sparsh1212 replied:
Done

BenHenning commented on scripts/src/java/org/oppia/android/scripts/xml/XmlSyntaxCheck.kt:
```diff
@@ -0,0 +1,85 @@
+  if (errorList.isNotEmpty()) {
+    errorList.forEach { error ->
+      val failureMessage =
+        "${error.getSystemId().removePrefix("file:")}:${error.getLineNumber()}:${error.getColumnNumber()}: ${error.message}"
```
Hmm parsing the system ID like this seems a bit hacky. I wonder if we could instead create a separate error handler for each file (by using a separate document builder) so that we know exactly which file corresponds to each error? While a bit more cumbersome, that relies less on an identifier happening to include the filename (which sounds like something that could potentially change in the future).

Sparsh1212 replied:
Done

BenHenning commented on scripts/src/java/org/oppia/android/scripts/xml/XmlSyntaxErrorHandler.kt:
```diff
@@ -0,0 +1,33 @@
+import org.xml.sax.SAXParseException
+
+/**
+ * Class for custom error handling of the parse exceptions thrown by the parser. It collects all
+ * the syntax errors thrown by the parser, which can then later be retrieved.
```
Could be made a bit more concise I think:

```suggestion
 * Custom XML parser error handler which collects syntax errors as they occur for later processing.
```

Sparsh1212 replied:
Done

BenHenning commented on scripts/src/java/org/oppia/android/scripts/xml/XmlSyntaxErrorHandler.kt:
```diff
@@ -0,0 +1,33 @@
+  /**
+   * Retrieves all the errors collected by the handler.
+   *
+   * @return List<SAXParseException> a list of all the errors collected by the error handler
```
```suggestion
   * @return a list of all the errors collected by the error handler
```

We don't specify the type in `@return` clauses.

Sparsh1212 replied:
Done

BenHenning commented on scripts/src/java/org/oppia/android/scripts/xml/XmlSyntaxErrorHandlerTest.kt:
```diff
@@ -0,0 +1,147 @@
+    parseXml(docBuilder = docBuilder, file = tempFile)
+
+    val errorList = xmlSyntaxErrorHandler.retrieveErrorList()
+    assertThat(errorList.isEmpty()).isEqualTo(true)
```
Prefer:

```kotlin
assertThat(errorList).isEmpty()
```

Sparsh1212 replied:
Done

BenHenning commented on scripts/src/java/org/oppia/android/scripts/xml/XmlSyntaxErrorHandlerTest.kt:
```diff
@@ -0,0 +1,147 @@
+
+    val errorList = xmlSyntaxErrorHandler.retrieveErrorList()
+    assertThat(errorList.size).isEqualTo(1)
+    assertThat(errorList.elementAt(0).message).isEqualTo(
```
Bit more Kotlin-idiomatic to do ``errorList.first()`` or ``errorList.single()`` in these cases.

Sparsh1212 replied:
Done

BenHenning commented on scripts/src/java/org/oppia/android/scripts/xml/XmlSyntaxErrorHandlerTest.kt:
```diff
@@ -0,0 +1,147 @@
+  }
+
+  @Test
+  fun testXmlErrorHandler_multipleinvalidXmlFiles_errorHandlerShouldCollectAllErrors() {
```
```suggestion
  fun testXmlErrorHandler_multipleInvalidXmlFiles_errorHandlerShouldCollectAllErrors() {
```

Typo in test name.

Sparsh1212 replied:
This is outdated, as we are using a separate error handler for every file, so we are not using the same handler for multiple files.

BenHenning reviewed (commented):
Thanks @Sparsh1212. Took another pass on the PR.

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 commented:
Thanks, @BenHenning.
Changes done.

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
bazel files nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Introduce a library for the regex assets

BenHenning commented:
@Sparsh1212 please assign back to me once this PR's chain (& subsequently this PR) has been updated to be based off of #3374.

Sparsh1212 committed:
Make directory structure same as that of #3374

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Make directory structure similar to #3374

Sparsh1212 commented:
@BenHenning This PR's directory structure is now updated based on #3374.
PTAL.

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
merge develop

Sparsh1212 committed:
diable ktlint max-line-length

Sparsh1212 committed:
disable ktlint-max-line

Sparsh1212 committed:
disable ktlint max-length

Sparsh1212 committed:
merge regex-checks

Sparsh1212 committed:
remove testonly attribute from tests

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

BenHenning commented on scripts/src/java/org/oppia/android/scripts/xml/BUILD.bazel:
```diff
@@ -0,0 +1,25 @@
+    name = "xml_syntax_error_handler_lib",
+    testonly = True,
+    srcs = ["XmlSyntaxErrorHandler.kt"],
+    visibility = ["//scripts:oppia_script_binary_visibility"],
```
Could this be ``oppia_script_test_visibility`` instead?

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/xml/XmlSyntaxCheck.kt:
```diff
@@ -0,0 +1,92 @@
+    val fileErrorList = xmlSyntaxErrorHandler.retrieveErrorList()
+    if (fileErrorList.isNotEmpty()) {
+      fileErrorList.forEach { error ->
+        allErrorsList.add(Pair(error, file))
```
It'd be more Kotlin-idiomatic to compute the errors list as a result of this overall expression (similar to how we use fold/reduce/map/flatMap elsewhere). I suggest trying to rearrange this logic to compute the resulting list rather than mutating it within lambdas.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/xml/XmlSyntaxCheckTest.kt:
```diff
@@ -0,0 +1,128 @@
+    assertThat(exception).hasMessageThat().contains(XML_SYNTAX_CHECK_FAILED_OUTPUT_INDICATOR)
+    assertThat(outContent.toString().trim()).isEqualTo(
+      """
+      ${retrieveTestFilesDirectoryPath()}/TestFile.xml:6:8: The end-tag for element type "shape" must end with a '>' delimiter.
```
Here & elsewhere: we should try to avoid exceeding the column limit, and in this case I think it's avoidable. Consider changing this to:

```kotlin
assertThat(outContent.toString()).contains("${retrieveTestFilesDirectoryPath()}/TestFile1.xml:6:8: The end-tag")
assertThat(outContent.toString()).contains("The end-tag for element type "shape" must end with a '>' delimiter.")
```

In this case I don't think it's important to necessarily verify the exact error since that's dependent on the XML parser, but it is important to verify the formatting & presence of the error since that's influenced by code we maintain.

Sparsh1212 replied:
Revamped the tests to check output format, pass status and failures.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/xml/XmlSyntaxErrorHandlerTest.kt:
```diff
@@ -0,0 +1,88 @@
+    parseXml(docBuilder = docBuilder, file = tempFile)
+
+    val errorList = xmlSyntaxErrorHandler.retrieveErrorList()
+    assertThat(errorList.size).isEqualTo(1)
```
```suggestion
    assertThat(errorList.size).hasSize(1)
```

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/xml/XmlSyntaxErrorHandlerTest.kt:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  @Test
+  fun testXmlErrorHandler_invalidXml_errorListShouldBeEmpty() {
```
Suggest adding tests for multiple errors, and for warning/error/fatal error cases (consider that if any lines of code in XmlSyntaxErrorHandler are commented out then at least 1 test should fail).

Sparsh1212 replied:
Ah! I just dug a bit deep into this and realized that, for our use case, only `fatalError` is essential to us. Because, whenever there is any syntactically incorrect XML present, then always a `fatalError` is thrown, and for this case as we are only concerned of the syntax error hence, all our errors in testing will be of type `fatalError` only. 
Also, there won't be the case of multiple failures in a single XML file because, as soon as the parser encounters a `fatalError` it stops the further processing of XML.
Hence, it seems like we won't need tests for multiple failures as we will never get more than a single error and also the use cases of `warning` and `error` are irrelevant for our case. 
I was probably thinking of removing them from the syntax error handler but it is compulsory to override them.

Sparsh1212 replied:
Added tests for all cases. PTAL.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/xml/XmlSyntaxErrorHandlerTest.kt:
```diff
@@ -0,0 +1,88 @@
+  private fun parseXml(docBuilder: DocumentBuilder, file: File) {
+    try {
+      docBuilder.parse(file)
+    } catch (e: SAXParseException) { }
```
I suggest adding a comment both here within the catch block & in the production version to explain why it's okay & correct to ignore the caught exception.

Sparsh1212 replied:
Added.

BenHenning reviewed (changes requested):
Thanks @Sparsh1212! PR generally LGTM--I think the latest iteration of the checker is really clear & easy to follow. My main concern is that we need to verify the new tests are passing in CI which is blocked on the issues #3340 is running into.

Sparsh1212 committed:
Apply review suggestions

Sparsh1212 committed:
ktlint fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Apply nit fixes

Sparsh1212 commented:
> Thanks @Sparsh1212! PR generally LGTM--I think the latest iteration of the checker is really clear & easy to follow. My main concern is that we need to verify the new tests are passing in CI which is blocked on the issues #3340 is running into.

Addressed all the nits. PTAL.

Sparsh1212 committed:
Remove script constants

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
add tests for XML syntax error handler

Sparsh1212 committed:
add todo

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/xml/XmlSyntaxErrorHandlerTest.kt:
```diff
@@ -0,0 +1,150 @@
+  }
+
+  @Test
+  fun testXmlErrorHandler_invokeErrorCase_errorShouldBeCollected() {
```
It feels a bit like fatal error is missing here. Maybe add a separate test for it even though it's technically covered in parsing test above (consider that implementation changes may result in the above error no longer being a fatal error which technically would reduce behavioral coverage in this test suite).

Sparsh1212 replied:
Added.

BenHenning reviewed (commented):
Thanks @Sparsh1212! Same as #3340. This is effectively 'LGTM' with one more comment to resolve. Please resolve that & assign back once all CI tests are passing in this PR.

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
add test

Sparsh1212 committed:
add testOnly

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
nit fix

BenHenning reviewed (commented):
Thanks @Sparsh1212! LGTM, but waiting to approve until #3340 is merged. Also, per our discussion I'm cancelling the workflows for this & the next PR since that will relieve pressure on the CI system and once you merge #3340 into develop and update these PRs they should have much fewer tests to run..

Sparsh1212 committed:
merge-regex-checks

BenHenning reviewed (approved):
Thanks @Sparsh1212! CI is green so going ahead and merging this.
